File: dedtmf.m
function [Y,ER,Fs,Rs,Ts,Bfe,Afe] = dedtmf(X,P,W,H,params)
% [Y,ER,Fs,Rs,Ts,Bf,Af] = dedtmf(X,P,W,H,params)
%   Remove DTMF (and other stationary tones) from signal
%   Inputs:
%    X  - input waveform
%    P  - order of LPC models to fit 
%    W  - length of blocks for LPC fitting
%    H  - hop between successive blocks
%    params - additional parameters:
%     params.poleradthreash (0.98) - radius threshold for mapping poles to zeros
%     params.poleradtrans (0.002) - radius transition width for mapping
%     params.polerad (0.98) - radius for added compensatory poles
%     params.noiseexcite (0) - use white noise excitation (to debug)
%   Returns:
%    Y  - audio filtered to remove steady tones
%    ER - block-wise energy ratio of original to filtered (close to
%         zero where significant tones have been removed.
%    Fs, Rs, Ts - frequencies (0..pi), radii (0..1) and times (in
%         samples) of the main components found.
%    Bf, Af - per-frame coefficients of the tone-removal filters used.
%
% 2013-11-22 Dan Ellis dpwe@ee.columbia.edu

if nargin < 2; P = 16; end
if nargin < 3; W = 1024; end
if nargin < 4; H = W/2; end
if nargin < 5; params.dummy = []; end  % to make it a struct

VERSION=0.1;
DATE=20131126;

% default parameters
if isfield(params, 'poleradthresh') == 0; params.poleradthresh = 0.98; end
if isfield(params, 'poleradtrans') == 0; params.poleradtrans = 0.002; end
if isfield(params, 'polerad') == 0; params.polerad = 0.98; end
if isfield(params, 'noiseexcite') == 0; params.noiseexcite = 0; end

% Break x into w-length blocks stepped by h
Xf = frame(X,W,H);

% Fit LPC models to every frame
Af = lpc(Xf, P);

% Find all the poles of the per-frame LPC filters
Rf = rootsbyrow(Af);
% Magnitudes of each pole
Mf = abs(Rf);
% Corresponding pure-phase parts
Cf = Rf./Mf;

% Build times, magnitudes, and angles of all positive freq roots 
% to return in diagnostics
Ts = repmat(W/2+H*[0:size(Mf,1)-1]',1,size(Mf,2));
Ts = Ts(:);
Rs = Mf(:);
Fs = angle(Cf(:));
% Keep just +ve freq complex poles
keep = find(Fs>0 & Fs <pi);
Ts = Ts(keep);
Rs = Rs(keep);
Fs = Fs(keep);

% Modify magnitude; ensure smaller than 1
% K = 0.25;
%Mfm = (1-abs(1-Mf)).^K;
% Use a sigmoid to make poles close to 1 even closer
%Mfm = (Mf > poleradthresh);
Mfm = sigmoid( (Mf - params.poleradthresh)/params.poleradtrans );
% New polynomial with exaggerated pole radii
Bfe = polybyrow( Mfm .* Cf );
% Compensatory poles - just inside zeros
Afe = polybyrow( params.polerad*Mfm .* Cf);

% Apply inverse filters to each block of X
if params.noiseexcite
  % Use white noise excitation of output to debug filtering
  E = randn(length(X), 1);
  Xf = frame(E,W,H);
end
[nr,nc] = size(Xf);
Xff = zeros(nr, nc);
for i = 1:nc
  Xff(:,i) = filter(Bfe(i,:),Afe(i,:),Xf(:,i)')';
end

% Reconstruct with 50%-overlapped hanning window
L = 2 * H;
pad = W - L;
win = [zeros(1,floor(pad/2)), hanning(L)', zeros(1,ceil(pad/2))];

% Figure per-block energy ratio
ER = sum(bsxfun(@times, Xff, win').^2)./sum(bsxfun(@times, Xf, win').^2);
% Threshold
ER = min(ER, 1.0);
% scale
%Xff = repmat(ER, size(Xff,1), 1).* Xff;

%Y = ola(diag(win)*Xff, H);
Y = ola(bsxfun(@times, Xff, win'), H);
%Y = ola(diag(win)*Xf, H);



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function R = rootsbyrow(A)
% Calculate the roots for the polynomials on each row of A
[nr,nc] = size(A);
R = zeros(nr, nc-1);
for i = 1:nr
  R(i,:) = roots(A(i,:))';
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function A = polybyrow(R)
% Calculate the polynomial coefficients for rows of roots in R
[nr,nc] = size(R);
A = zeros(nr, nc+1);
for i = 1:nr
  A(i,:) = poly(R(i,:)');
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function Y = sigmoid(X)
Y = 1./(1+exp(-X));

File: dedtmf.py
#!/usr/bin/env python
"""
dedtmf.py

Python port of the dtmf-tone removal matlab code.

2014-01-23 Dan Ellis dpwe@ee.columbia.edu
"""

import numpy as np
# For filter
import scipy.signal
# For SRI's wavreading code
import scipy.io.wavfile as wav
# For command line
import sys

def frame(x, window, hop):
    """ Convert vector x into an array of successive window-point segments, 
        stepped by hop
        Done with stride_tricks, no copying, but have to lose the final 
        part-window 
    """

    nframes = 1 + int( np.floor( (len(x)-window)/hop ))
    shape = (nframes, window)
    strides = (x.strides[0] * hop, x.strides[0])
    return np.lib.stride_tricks.as_strided(x, shape=shape, strides=strides)

def ola(X, hop):
    """ Overlap-add rows of X into a single vector, stepping by hop """
    nw, W = np.shape(X)

    # How long should X be in the end?
    # one complete window, plus another hop's worth of points for each
    # additional window
    lenx = W + (nw-1)*hop
    Y = np.zeros(lenx)

    for i in range(nw):
        Y[i*hop : i*hop+W] += X[i,]

    return Y

# based on https://code.google.com/p/vocoder-lpc/source/browse/lpc.py

def autoco(x, p):
    """
    Calculate autocorrelation of x
    Return only the first p values, i.e. r_xx[0] to r_xx[p-1]
    """
    # Naive version
    #R = np.correlate(x, x, 'full')
    #R = R[len(x)-1:len(x)+p]
    # Better version (> twice as fast?)
    X = np.fft.rfft(np.r_[x, np.zeros(p-1)])
    rxx = np.fft.irfft(np.square(np.abs(X)))
    return rxx[:p]

def lpc_frame(x, p):
    """
    Fit p'th order LPC model to a single frame of data in x
    using Levinson-Durbin recursion
    """
    R = autoco(x, p+1)
    a = np.zeros((p+1, p+1))
    k = np.zeros(p+1)

    E = np.r_[R[0], np.zeros(p)]

    fa = np.zeros(p+1)
    fa[0] = 1

    if R[0] > 0:
      # Only do the recursion if this frame is nonempty
      for i in xrange(1, p + 1):
        c = 0
        for j in xrange(1, i):
           c += a[j, i-1] * R[i-j]
        k[i] = (R[i] - c) / E[i-1]

        a[i][i] = k[i]

        for j in xrange(1, i):
           a[j][i] = a[j][i-1] - k[i] * a[i-j][i-1]

        E[i] = (1 - k[i]**2) * E[i-1]

      for j in xrange(1, p+1):
        fa[j] = -a[j][p]

    return fa, E[p]

def lpc(X, order):
    """ Calculate LPC fit to each row of X, return as rows of A matrix, 
        plus vector of energies
    """
    rows, cols = np.shape(X)
    A = np.zeros( (rows, order+1) )
    E = np.zeros(rows)
    for row in range(rows):
        #print "frame ", row, " of ", rows
        A[row,], E[row] = lpc_frame(X[row,:], order)
        if E[row] == 0:
            print "Empty frame: ", row, " of ", rows
    return A, E


def rootsbyrow(A):
    """ Calculate the roots for polynomials defined by each row of A
    """
    nr, nc = np.shape(A)
    R = np.zeros( (nr, nc-1), complex )
    for i in range(nr):
        R[i,] = np.roots(A[i,])
    return R

def polybyrow(R):
    """ Calculate the polynomial coefficients for rows of roots in R
    """
    nr, nc = np.shape(R)
    A = np.zeros((nr, nc+1))
    for i in range(nr):
        A[i,] = np.real(np.poly(R[i,]))
    return A

def sigmoid(X):
    """ simple sigmoid function """
    return 1./(1.+np.exp(-X))

########## main dedtmf ############

def dedtmf(X, lpc_order=40, win_pts=4096, hop_pts=256, params={}):
    """ Remove sustained, steady tones from an audio signal by finding strong 
        poles in LPC analysis on long-ish blocks, then inverse-filtering to 
        remove them.
    """
    # Set up default params
    # poleradthresh is the threshold for "capturing" poles
    if 'poleradthresh' in params:
        poleradthresh = params['poleradthresh']
    else:
        poleradthresh = 0.98
    # poleradtrans is the transition width for pole capture
    if 'poleradtrans' in params:
        poleradtrans = params['poleradtrans']
    else:
        poleradtrans = 0.002
    # polerad is the radius of the new poles put behind the introduced 
    # circles to limit the notch width
    if 'polerad' in params:
        polerad = params['polerad']
    else:
        polerad = 0.98

    # Default hop is half whatever window is
    if hop_pts == 0:
        hop_pts = win_pts/2

    # Break x into w-length blocks stepped by h
    Xf = frame(X, win_pts, hop_pts)

    print "lpc fitting frames..."
    # Fit LPC models to every frame
    Af, Ef = lpc(Xf, lpc_order)

    # Find all the poles of the per-frame LPC filters
    print "finding roots..."
    Rf = rootsbyrow(Af);
    # Magnitudes of each pole
    Mf = np.absolute(Rf)
    # Corresponding pure-phase parts
    Cf = Rf/(Mf + (Mf==0))

    # Modify magnitude; ensure smaller than 1
    # Use a sigmoid to make poles close to 1 even closer
    # (defaults now set at top of function)
    #poleradthresh = 0.98
    #poleradtrans =  0.002
    #Mfm = (Mf > poleradthresh);
    Mfm = sigmoid( (Mf - poleradthresh)/poleradtrans )
    # New polynomial with exaggerated pole radii
    Bfe = polybyrow( Mfm * Cf )
    # Compensatory poles - just inside zeros
    #polerad = 0.98
    Afe = polybyrow( polerad*Mfm * Cf);

    # Apply inverse filters to each block of X
    print "applying filters ..."
    nr, nc = np.shape(Xf)
    Xff = np.zeros( (nr, nc) )
    for i in range(nr):
        Xff[i,] = scipy.signal.lfilter(Bfe[i,], Afe[i,], Xf[i,])

    # Reconstruct with 50%-overlapped hanning window
    L = 2 * hop_pts
    pad = win_pts - L
    # Was an L point hanning, this is to make it flat
    win = np.r_[np.zeros(int(np.floor(pad/2))), 
                np.hanning(L+1),  
                np.zeros(int(np.ceil(pad/2)-1))]

    # Figure per-block energy ratio
    #ER = (np.sum(np.power(Xff*win, 2.0), axis=-1) /
    #      np.sum(np.power(Xf*win, 2.0), axis=-1))
    # Threshold
    #ER = np.minimum(ER, 1.0);
    # scale
    #Xff = repmat(ER, size(Xff,1), 1).* Xff;
    
    print "overlap-add ..."
    return ola(Xff*win, hop_pts)


############## Provide a command-line wrapper

def to_num(s):
    ''' Converts a string to an int or a float if that works, 
        else leaves it as a string'''
    try:
        return int(s)
    except ValueError:
        try:
            return float(s)
        except ValueError:
            return s

def parse_param_file(filename, defaults):
    """ Read a param file of "key value" lines.  Return a dictionary of 
        assignments.  Fill in defaults from default dict """
    result = defaults
    if len(filename):
        with open(filename, 'r') as f:
            for line in f:
                keyval = line.lstrip().split(None, 1)
                if len(keyval) > 0:
                    key = keyval[0]
                    if len(keyval) > 1:
                        val = keyval[1]
                    else:
                        val = []
                    if key != "" and key[0] != '#':
                        # Overwrite or create
                        result[key] = to_num(val)
    return result

def main(argv):
    """ Main routine to apply dedtmf from command line """
    if len(argv) != 3 and len(argv) != 4:
        raise NameError( ("Usage: ", argv[0], 
                          " inputsound.wav outsound.wav [paramfile]") )

    # Input and output files
    inwavfile = argv[1]
    outwavfile = argv[2]

    # Read in param file (if any) and/or setup default values
    paramfile = ''
    if len(argv) == 4:
        paramfile = argv[3]
    params = parse_param_file(paramfile, 
                              {'win_pts':4096, 
                               'hop_pts':256, 
                               'lpc_order':40, 
                               'poleradthresh':0.98, 
                               'poleradtrans':0.002, 
                               'polerad':0.98})
    #print params

    # Read in wav file
    srate, wavd = wav.read(inwavfile)
    # normalize short ints to floats of -1 / 1
    data = np.asfarray(wavd) / 32768.0  

    lendata = len(data)
    #win_pts = 4096
    #hop_pts = 256
    win_pts = params['win_pts']
    hop_pts = params['hop_pts']
    # Pad with zeros - half a window before, a whole window after 
    # just to make sure dropping the final part-window doesn't lose anything 
    prepad_pts = win_pts/2
    pdata = np.r_[np.zeros(prepad_pts), data, np.zeros(win_pts)]

    # Apply
    #lpc_order = 40
    lpc_order = params['lpc_order']
    filtrd = dedtmf(pdata, lpc_order, win_pts, hop_pts, params)

    # Strip the padding
    udata = filtrd[prepad_pts:prepad_pts + lendata]

    # Write the wav out
    wav.write(outwavfile, srate, (32768.0*udata).astype(np.int16))
    print "Wrote ", outwavfile

# Actually run main
if __name__ == "__main__":
    main(sys.argv)



